### ISSUE TITLE
question (no an issue): where should I look to add a "move to trash" feature?

### ISSUE THREAD

**Issue by hradec** [zevv/duc#248]
I'm evaluating duc as not only a tool to show disk usage on network storage, but also as a frontend to show disk usage of projects, where users would have the option to flag high-usage folders to be deleted (like moving to the trash for later deletion).

The ideia is a simple and quick frontend to encourage artists in a VFX studio to do disk cleanup, since they are the ones that known what can and cannot be deleted. 

I was wondering if someone could point me out to where should I look in the source to implement such feature, if it's not too much trouble, just so I could get a head start on prototyping it. 

anyhow, thanks for this wonderful tool!!

**Comment by l8gravely** [comment]
Hradec> I'm evaluating duc as not only a tool to show disk usage on
Hradec> network storage, but also as a frontend to show disk usage of
Hradec> projects, where users would have the option to flag high-usage
Hradec> folders to be deleted (like moving to the trash for later
Hradec> deletion).

duc isn't really an interactive tool like this.  But maybe there could
be a way to generate an output file which could be fed into another
tool (xargs, etc) to do the actual work.  

Hradec> The idea is a simple and quick frontend to encourage artists
Hradec> in a VFX studio to do disk cleanup, since they are the ones
Hradec> that known what can and cannot be deleted.

Just having the web pages up has been a good way for me to get my end
users to do some cleanup.  Wait, what am I saying?  End users *never*
cleanup unless you hold a gun to their head.  It's a law of nature.
*grin*

Hradec> I was wondering if someone could point me out to where should
Hradec> I look in the source to implement such feature, if it's not
Hradec> too much trouble, just so I could get a head start on
Hradec> prototyping it.

I'd probably start looking in src/duc/cmd-ui.c for the ncurses
interface since it might make the most sense to start there.  cmd-gui
is the graphical interface, which I've honestly never used.  

The next file would be the cmd-cgi.c in the same directory, you might
be able to put in some javascript to let you mark a directory to
purge.  Not sure how it would work, would have to poke at it.
Especially since you don't want the CGI version writing random files
on the web server, which might not even be accessible to your end
users beyond HTTP.  

How do you present the data now to your users?  I suspec the web
interface, so that might be the place to start.

But honestly, it all boils down to a management issue.  duc just
provides the tool to see where the problem is, not to actually fix it.

I even have other tools I use to find and list the top N files in
terms of size in a directory tree.  Finding that single pair of 800gb
files that the user didn't realize were being created was a big help.

And setting per-user disk quotas, even if only for reporting purposes,
is a good thing as well.  Just being able to generate a weekly report
saying "Bob is using X, and Charlie is using X/2", and so forth means
that Bob and or his manager might have some incentive to do cleanup.

But anyway, look at cmd-ui.c for a first stab maybe.

John

**Comment by aspiers** [comment]
[**@l8gravely**](https://github.com/l8gravely) commented on [June 29, 2020 1:08 AM](https://github.com/zevv/duc/issues/248#issuecomment-650841860 "2020-06-29T00:08:09Z - Replied by Github Reply Comments"):
> duc isn't really an interactive tool like this.

I'm a bit confused by this comment, as I've just discovered duc and it seems *awesomely* interactive to me ;-) (as well as awesome in many other ways).  It's so great that it has not one, not two, but *three* interactive interfaces.  That puts it way ahead of the competition in this respect.

Presumably the vast majority of use cases are when disk space needs to be freed, so it makes a huge amount of sense to me for a disk usage utility to also support deletion.  Big +1 for this feature request!

**Comment by zevv** [comment]
Quoting Adam Spiers (2020-09-26 17:39:02)
> [**@l8gravely**](https://github.com/l8gravely) commented on [June 29, 2020 1:08 AM](https://github.com/zevv/duc/issues/248#issuecomment-650841860 "2020-06-29T00:08:09Z - Replied by Github Reply Comments"):
> > duc isn't really an interactive tool like this.
> 
> I'm a bit confused by this comment, as I've just discovered duc and it
> seems *awesomely* interactive to me ;-) (as well as awesome in many
> other ways).  It's so great that it has not one, not two, but *three*
> interactive interfaces.  That puts it way ahead of the competition in
> this respect.

It's a design philosophy thing. Duc is pretty flexible and allows a lot
of different use cases, one of which is pretty common: index on machine
A, and inspect the db on machine B. In this case, there is no relation
between the files in the database and the files on the disk, so 
operations on the files in the do not make sense. I don't think I would
be confortable putting this option in Duc, I wouldn't want to
responsible for deleting anyones data, especially on the wrong machine :)

--
:wq
^X^Cy^K^X^C^C^C^C

**Comment by aspiers** [comment]
Thanks a lot for the quick reply!

I totally appreciate the design decision to decouple indexing from analysis.  That's a really really nice feature.  However, it feels a little like throwing the baby out with the bath water to use this as a reason for not supporting deletion.  Surely a large proportion of users will be both indexing and analysing on the same machine, and in these cases, deletion is really useful.  For example, I'm using it to manage disk usage on some remote servers, running the ncurses interface over `ssh`.

Furthermore, it should be easy to add a safety check which ensures that deletion is only allowed on the machine where the index came from.  For example, it would be easy to take a hash of values unique to the machine being indexed, such as the fully qualified hostname, perhaps combined with a sorted list of MAC addresses, or something like that.  If anything doesn't match between the hash generated for the index and for the machine running the interface, then simply prohibit deletion.  And if the hostname or MAC addresses change, simply suggest to the user that if they're convinced it's the same machine, they probably need to re-index.

I sincerely believe this would provide safe deletion functionality to a large subset of users who could benefit from it.

**Comment by l8gravely** [comment]
>>>>> "Adam" == Adam Spiers <notifications@github.com> writes:

Adam> @l8gravely commented on June 29, 2020 1:08 AM:
Adam>     duc isn't really an interactive tool like this.
   
Adam> I'm a bit confused by this comment, as I've just discovered duc
Adam> and it seems awesomely interactive to me ;-) (as well as awesome
Adam> in many other ways). It's so great that it has not one, not two,
Adam> but three interactive interfaces. That puts it way ahead of the
Adam> competition in this respect.

That's just the tool which is looking at the generated database of
disk usage. You can't use duc to examine a filesystem.  It's like
'windirstat' in that regards, it needs to first scan the filesystem to
build up the data, then you can examine it interactively.  

Adam> Presumably the vast majority of use cases are when disk space
Adam> needs to be freed, so it makes a huge amount of sense to me for
Adam> a disk usage utility to also support deletion. Big +1 for this
Adam> feature request!

We will look into it, but... since the DB might be examined on a
system which isn't connected to the actual filesystem running out of
space (think the web GUI) that doesn't make sense.

Also, duc just lists directories and their sizes, not so much the
files in each directory, so again it's not really an interactive tool
which examines the filesystem in real time.

John

**Comment by aspiers** [comment]
[**@l8gravely**](https://github.com/l8gravely) commented on [September 26, 2020 7:22 PM](https://github.com/zevv/duc/issues/248#issuecomment-699530645 "2020-09-26T18:22:58Z - Replied by Github Reply Comments"):
> That's just the tool which is looking at the generated database of disk usage. You can't use duc to examine a filesystem. It's like 'windirstat' in that regards, it needs to first scan the filesystem to build up the data, then you can examine it interactively.

Sure, I appreciate that as noted above.

> We will look into it, but... since the DB might be examined on a system which isn't connected to the actual filesystem running out of space (think the web GUI) that doesn't make sense.

I agree it doesn't make sense when examined on a different system.  That's why I proposed the above solution which prevents that from happening.

> Also, duc just lists directories and their sizes, not so much the files in each directory, so again it's not really an interactive tool which examines the filesystem in real time.

I appreciate that the filesystem data doesn't currently get reindexed from the interactive interfaces, but they're still interactive.  And again, in the case where the interface is running on the same system from which the indexed data originates, there is real benefit in allowing not just deletion and automatic re-indexing of the containing directory, but also manually initiated re-indexing of any selected directory.  These features are available in "competitors" like qdirstat / kdirstat, and I find them extremely useful.

Thanks a lot for being willing to listen to feedback!

**Comment by l8gravely** [comment]
>>>>> "Adam" == Adam Spiers <notifications@github.com> writes:

Adam> @l8gravely commented on September 26, 2020 7:22 PM:

Adam>     That's just the tool which is looking at the generated
Adam>     database of disk usage. You can't use duc to examine a
Adam>     filesystem. It's like 'windirstat' in that regards, it needs
Adam>     to first scan the filesystem to build up the data, then you
Adam>     can examine it interactively.
   
Adam> Sure, I appreciate that as noted above.

Adam>     We will look into it, but... since the DB might be examined
Adam>     on a system which isn't connected to the actual filesystem
Adam>     running out of space (think the web GUI) that doesn't make
Adam>     sense.
   
Adam> I agree it doesn't make sense when examined on a different
Adam> system. That's why I proposed the above solution which prevents
Adam> that from happening.

Adam>     Also, duc just lists directories and their sizes, not so
Adam>     much the files in each directory, so again it's not really
Adam>     an interactive tool which examines the filesystem in real
Adam>     time.
   
Adam> I appreciate that the filesystem data doesn't currently get
Adam> reindexed from the interactive interfaces, but they're still
Adam> interactive. And again, in the case where the interface is
Adam> running on the same system from which the indexed data
Adam> originates, there is real benefit in allowing not just deletion
Adam> and automatic re-indexing of the containing directory, but also
Adam> manually initiated re-indexing of any selected directory. These
Adam> features are available in "competitors" like qdirstat /
Adam> kdirstat, and I find them extremely useful.

Then you should use those tools instead, since they cover a use case
which duc doesn't handle in a way that fits your needs.  

Adam> Thanks a lot for being willing to listen to feedback!

We are, but you need to be willing to listen to us as well.  duc (and
it's previous ruby based version philesight) were designed for *large*
filesystems, where an interactive 'du' just didn't work, because it
took too long to return useful results.

When you have millions of files and directories, and Terabytes of
storgage, windirstat, qdirstat, du, etc just do not work.  They don't
scale properly.  The indexing process along can take hours.  This is
why duc is split into two seperate phases.  The first indexes, the
second is to explore.

Even though windirstat let's you delete a file from within it and then
does re-indexing, it doesn't update instantly.  It still needs to
re-scan everything.  This is one thing when you do it on a laptop SSD,
but trying doing it on a 15Tb filesystem with 10 million files and
directories.

So I'm not going to work on this feature, but please do try to work on
it yourself!  If you can come up with a working patch, we will
consider applying it to the code base and offering it as an option.
But it's not a simple problem.

Cheers,
John

**Comment by aspiers** [comment]
[**@l8gravely**](https://github.com/l8gravely) commented on [September 27, 2020 5:23 PM](https://github.com/zevv/duc/issues/248#issuecomment-699656047 "2020-09-27T16:23:46Z - Replied by Github Reply Comments"):
> duc (and it's previous ruby based version philesight) were designed for \*large\* filesystems, where an interactive 'du' just didn't work, because it took too long to return useful results. When you have millions of files and directories, and Terabytes of storgage, windirstat, qdirstat, du, etc just do not work. They don't scale properly. The indexing process along can take hours.

Agreed!  That's exactly why I want to use duc instead of these alternatives ;-)

> This is why duc is split into two seperate phases. The first indexes, the second is to explore. Even though windirstat let's you delete a file from within it and then does re-indexing, it doesn't update instantly. It still needs to re-scan everything.

It's years since I've used windirstat, but in qdirstat and kdirstat, you can right-click on a directory and then choose an option "Refresh Selected", which only re-indexes that directory.  Also, this targeted re-indexing happens automatically to the containing directory after deleting something in it.  That design gives the best of both worlds: it allows the main indexing to be done in a separate phase, but can selectively update subtrees as appropriate, wrapping up the combination in a convenient interface.  IMHO it would be a significant boost to duc if it supported the same functionality in the specific case where the interface is being run on the same machine as the filesystem being indexed.

> This is one thing when you do it on a laptop SSD, but trying doing it on a 15Tb filesystem with 10 million files and directories.

Absolutely agree, which is why I was definitely not proposing to re-scan everything.  duc already allows selective re-indexing of subdirectories, so my proposal is simply to take advantage of that.

> So I'm not going to work on this feature, but please do try to work on it yourself! If you can come up with a working patch, we will consider applying it to the code base and offering it as an option.

Thanks for being open to this.  If it's simple then maybe I can give it a go.

> But it's not a simple problem.

Then I'm probably missing something, because it seems extremely simple to me:

1. Write some code which runs in all the UIs and figures out whether it's running on the same machine which has the indexed filesystem.  As mentioned above, this could simply calculate a hash of things like the hostname / MAC addresses to produce a uniquely identifying value.
2. If that check run within a UI succeeds, expand the UI to offer two new interactive operations:
    1. Re-index the selected directory.  This covers the case where the user knows that some external process has changed the disk usage within that directory (e.g. if they've manually deleted files themself via a file manager or the terminal, or perhaps if a rogue process is spewing data to sub-directories at a high rate and they want to figure out what is chewing up disk space).
    2. Delete the selected file or directory, and then automatically re-index the containing directory. 

Are any of my assumptions off base here?  Thanks!

**Comment by l8gravely** [comment]
>>>>> "Adam" == Adam Spiers <notifications@github.com> writes:

John> @l8gravely commented on September 27, 2020 5:23 PM:

John> duc (and it's previous ruby based version philesight) were
John> designed for *large* filesystems, where an interactive 'du' just
John> didn't work, because it took too long to return useful results.
John> When you have millions of files and directories, and Terabytes
John> of storgage, windirstat, qdirstat, du, etc just do not
John> work. They don't scale properly. The indexing process along can
John> take hours.
   
Adam> Agreed! That's exactly why I want to use duc instead of these
Adam> alternatives ;-)

You miss my point, *large* filesystems. I'm talking tens of millions
of files and terabytes of storage.  I'm not talking about a home
system with 1-2 million files.  Even then, kdirstat will take an
appreciable time to index, and then re-index files.

John> This is why duc is split into two seperate phases. The first
John> indexes, the second is to explore. Even though windirstat let's
John> you delete a file from within it and then does re-indexing, it
John> doesn't update instantly. It still needs to re-scan everything.
   
Adam> It's years since I've used windirstat, but in qdirstat and
Adam> kdirstat, you can right-click on a directory and then choose an
Adam> option "Refresh Selected", which only re-indexes that directory.
Adam> Also, this targeted re-indexing happens automatically to the
Adam> containing directory after deleting something in it. That design
Adam> gives the best of both worlds: it allows the main indexing to be
Adam> done in a separate phase, but can selectively update subtrees as
Adam> appropriate, wrapping up the combination in a convenient
Adam> interface. IMHO it would be a significant boost to duc if it
Adam> supported the same functionality in the specific case where the
Adam> interface is being run on the same machine as the filesystem
Adam> being indexed.

John> This is one thing when you do it on a laptop SSD, but trying
John> doing it on a 15Tb filesystem with 10 million files and
John> directories.
   
Adam> Absolutely agree, which is why I was definitely not proposing to
Adam> re-scan everything. duc already allows selective re-indexing of
Adam> subdirectories, so my proposal is simply to take advantage of
Adam> that.

John> So I'm not going to work on this feature, but please do try to
John> work on it yourself! If you can come up with a working patch, we
John> will consider applying it to the code base and offering it as an
John> option.
   
Adam> Thanks for being open to this. If it's simple then maybe I can give it a go.

Adam>     But it's not a simple problem.
   
Adam> Then I'm probably missing something, because it seems extremely
Adam> simple to me:

Adam>  1. Write some code which runs in all the UIs and figures out
Adam>     whether it's running on the same machine which has the
Adam>     indexed filesystem. As mentioned above, this could simply
Adam>     calculate a hash of things like the hostname / MAC addresses
Adam>     to produce a uniquely identifying value.

Not an easy thing to know.  Also, duc is designed to work in two
phases only.  First you index, then you browse with the ncurses
interface, the GUI, the CGI or just printing info to stdout.

It would take a re-architecting of duc to be interactive in the way
you want.  Have you looked at baobob instead?  Or maybe look at this
qdirstat github issue:

https://github.com/shundhammer/qdirstat/issues/97

And compare how things are done. 

Adam>  2. If that check run within a UI succeeds, expand the UI to offer two new interactive operations:

Adam>      1. Re-index the selected directory. This covers the case
Adam>         where the user knows that some external process has
Adam>         changed the disk usage within that directory (e.g. if
Adam>         they've manually deleted files themself via a file
Adam>         manager or the terminal, or perhaps if a rogue process
Adam>         is spewing data to sub-directories at a high rate and
Adam>         they want to figure out what is chewing up disk space).

Adam>      2. Delete the selected file or directory, and then
Adam>         automatically re-index the containing directory.

It's not that simple.  I've been thinking about this on and off today
and I recall some issues with scanning a filesystem to update a DB
while at the same time examing in with the CGI.

It's been a long time since I investigated this, so maybe it's not a
problem any more.  But the way I do my indexing is to index to a new
DB, then move it into place of the old one.  This way users can
examine things while indexing happens late at night when the load on
the system is (hopefully) lower.

I also don't want to have 10 users all running indexes at the same
time, it's a waste of resources.  

Remember, duc was written to solve large filesystem usage problems,
where interactive 'du' just didn't work.  And where kdirstat and other
tools were just too slow and caused too much performance impact on
systems.

I've got 400Tb of filesystems that I index, with probably 500 million
files.  If I made duc be interactive, it would never finish the
initial scan!  Now do you see the issue?

This is *not* a simple problem space.  I've written scripts to use
quotas on my Netapps for reporting.  Great way to find out which user
is hogging space overall, and has large numbers of files.  But it
doesn't tell me where.

Running a plain 'du' or kdirstat also doesn't help because the
scanning phase takes so long.  Philesight allows me to say to my
users:  "Hey, you're using too much space, please dig through here
and do some cleanup."  And it's helped, because they can visually dig
into the system and see where the space needs to be purged.  Heck,
sometimes they find large log files they didn't know that had
created.  Getting rid of a pair of 800gb log files takes a while when
you have snapshots enabled...

You can delete them, but the disk space doesn't come back until you
expire the snapshots.  Which is a painful situation to be in when you
can't expire single files out of a snapshot.

So basically, if you want something interactive, please use an
interactive tool, which is not duc.  It's a data gathering and data
exploration tool.  Not a real time directory manipulation tool.

Now maybe we can generate a list of files to delete, which can be fed
to another tool to do the deletions.  Then you re-schedule the
indexing for later.  But doing it live isn't going to happen.

John

**Comment by aspiers** [comment]
On Wed, 30 Sep 2020 at 04:03, John <notifications@github.com> wrote:
> John> When you have millions of files and directories, and Terabytes
> John> of storage, windirstat, qdirstat, du, etc just do not
> John> work. They don't scale properly. The indexing process along can
> John> take hours.
>
> Adam> Agreed! That's exactly why I want to use duc instead of these
> Adam> alternatives ;-)
>
> You miss my point, *large* filesystems. I'm talking tens of millions
> of files and terabytes of storage. I'm not talking about a home
> system with 1-2 million files. Even then, kdirstat will take an
> appreciable time to index, and then re-index files.

Trust me, I fully appreciate this ;-)  I've worked with
enterprise-level storage for many years.  But I think maybe we have
been talking at cross-purposes - please see below.

> Have you looked at baobob instead? Or maybe look at this
> qdirstat github issue:
>
> https://github.com/shundhammer/qdirstat/issues/97
>
> And compare how things are done.

Thanks for the link.  I actually worked at the same company as the
qdirstat author for many years ;-)

My main complaint with most of the alternatives is that they don't
persist the results of the indexing to disk, so if you want to perform
a disk usage analysis over a longer period of time, you have to either
leave the app running, or restart it and suffer the tedious and
expensive reindexing.  In contrast, duc avoids that problem very
nicely, allowing you to persist the index for as long as you want.
However it's a double-edged sword, since then the index can gradually
become out of date.  My understanding was that duc already allows you
to selectively re-index subtrees, but maybe I misunderstood and that
will actually create a separate section in the database for the
subdirectory rather than updating the subdirectory in the original
index?  "duc info" seems to suggest this, but I'm unsure how the
internals operate.

> Adam> 2. If that check run within a UI succeeds, expand the UI to offer two new interactive operations:
>
> Adam> 1. Re-index the selected directory. This covers the case
> Adam> where the user knows that some external process has
> Adam> changed the disk usage within that directory (e.g. if
> Adam> they've manually deleted files themself via a file
> Adam> manager or the terminal, or perhaps if a rogue process
> Adam> is spewing data to sub-directories at a high rate and
> Adam> they want to figure out what is chewing up disk space).
>
> Adam> 2. Delete the selected file or directory, and then
> Adam> automatically re-index the containing directory.

I just wanted to check we are on the same page here: when I refer to
"selected directory" and "containing directory" above, I *don't* mean
the root of the original (and potentially huge) index.

Instead I mean just the selected sub-directories in which the user has
deleted files and/or requested a re-indexing.  For example, these
subdirectories could only have 20 files in, of which 10 turn out to be
huge files representing unnecessary bloat from a misconfigured or
buggy application.  After the user deletes those 10 files, the
subdirectory only has 10 files left, which could now total only a few
MB of usage, even though the total usage across the whole index is
100s of terabytes and 100s of millions of files.  So re-indexing that
subdirectory would only take a fraction of a second.

I'm beginning to suspect that I didn't make this obvious enough,
because several times you have been re-emphasising the cost of a full
indexing operation on a huge filesystem, which I fully acknowledge,
but is not really relevant to the functionality change I'm proposing.

If this is the cause of some misunderstanding then I apologise for not
being clearer!

> It's not that simple. I've been thinking about this on and off today
> and I recall some issues with scanning a filesystem to update a DB
> while at the same time examing in with the CGI.
>
> It's been a long time since I investigated this, so maybe it's not a
> problem any more. But the way I do my indexing is to index to a new
> DB, then move it into place of the old one. This way users can
> examine things while indexing happens late at night when the load on
> the system is (hopefully) lower.
>
> I also don't want to have 10 users all running indexes at the same
> time, it's a waste of resources.

OK yeah, good points.  However this doesn't sound like anything that
some simple locking mechanism wouldn't solve.  It could prevent more
than one indexing operation happening at any given time, and also
guard against conflicts between readers (UIs) and writers (indexers).

> Remember, duc was written to solve large filesystem usage problems,
> where interactive 'du' just didn't work. And where kdirstat and other
> tools were just too slow and caused too much performance impact on
> systems.
>
> I've got 400Tb of filesystems that I index, with probably 500 million
> files. If I made duc be interactive, it would never finish the
> initial scan! Now do you see the issue?

No, I don't.  Let's take this exact scenario as an example of the use
case I'm proposing:

1. I index 400TB / 500 million files (which obviously takes a long time).
2. I fire up a duc UI, and start looking for some "bloat" i.e.
subdirectories with undesirably high disk usage.
3. I identify a bloated directory and remove the bloat from that
directory.  I might also need to address the root cause of whatever is
causing the bloat, and that may take much longer than just removing
the bloat.

Now I repeat step 3 many more times.  This process could potentially
span hours or even days.

Each time I do step 3, the index gets a little bit further from the
truth. Quite possibly other areas of the filesystem may be changing
too - but at this point my main concern as a user is that the duc UI
is updating so that areas of bloat I've already dealt with aren't
cluttering the UI and obscuring other areas of bloat which I still
need to look at.  And obviously I don't want to re-index the whole
400TB because I only care about updating the bits of the index which
I've cleaned up myself.

My current understanding is that:

a) none of the alternatives to duc are good options, either due to
poor performance, or the lack of persistence of the index to disk
b) currently the way to do this with duc is to manually run "duc index
$DIR" after trimming bloat from each $DIR, and then restart the UI.

Is that correct?

My proposal based on these assumptions was to tweak the UIs to improve
the UX b) so that *in this specific case* where they are running on
the same machine, they can trigger the "duc index" command for the
selected *sub*directory, using some simple locking scheme to avoid the
issues you identify above, and then refresh the display from the
updated index.  Only a tiny number of files and directories need to be
re-indexed, rather than the 500 million required when building the
initial index.

However *if* my assumption about selective reindexing is *incorrect*,
then that renders my current UI proposal as currently unrealistic, so
my next question would be: how hard would it be to add support for
that selective re-indexing as a nice next step?

**Comment by shundhammer** [comment]
> My main complaint with most of the alternatives is that they don't persist the results of the indexing to disk, so if you want to perform a disk usage analysis over a longer period of time, you have to either leave the app running, or restart it and suffer the tedious and expensive reindexing.

Actually, QDirStat (and its predecessor KDirStat) can do that: You can write the current content to a _cache_ file, and you can read that file upon the next program start. Use _File_ -> _Write to Cache File_.

The question is, however, if that's really worthwhile. The difference was very drastic in the olden days when the Linux kernel was quite bad at caching directory contents; today, not so much. It depends on your usage pattern, of course: The very first time, all directories and i-node information (`stat()`) has to be read from disk; but the kernel caches that information up to a point where a subsequent read is near-instant.

Compare with the benchmarks I once did over several disk usage tools:

https://github.com/shundhammer/qdirstat/issues/97

See the _cold_ vs. _hot_ cache columns. Reading the information from a cache file isn't really quicker than that, and you have up-to-date information (the cache file is a snapshot from some time ago).

Having said that, QDirStat also comes with a Perl script that can create a cache file as a background / cron job; thus, you can even easily use it for a remote headless server, fetch that information and view it at your leisure on your desktop machine with QDirStat. See also

- https://github.com/shundhammer/qdirstat/blob/master/scripts/README.md
- https://github.com/shundhammer/qdirstat/blob/master/doc/QDirStat-for-Servers.md
- https://github.com/shundhammer/qdirstat/blob/master/doc/cache-file-format.txt

HTH

**Comment by shundhammer** [comment]
> My main complaint with most of the alternatives is that they don't persist the results of the indexing to disk, so if you want to perform a disk usage analysis over a longer period of time, you have to either leave the app running, or restart it and suffer the tedious and expensive reindexing.

Actually, QDirStat (and its predecessor KDirStat) can do that: You can write the current content to a _cache_ file, and you can read that file upon the next program start. Use _File_ -> _Write to Cache File_.

The question is, however, if that's really worthwhile. The difference was very drastic in the olden days when the Linux kernel was quite bad at caching directory contents; today, not so much. It depends on your usage pattern, of course: The very first time, all directories and i-node information (`stat()`) has to be read from disk; but the kernel caches that information up to a point where a subsequent read is near-instant.

Compare with the benchmarks I once did over several disk usage tools:

https://github.com/shundhammer/qdirstat/issues/97

See the _cold_ vs. _hot_ cache columns. Reading the information from a cache file isn't really quicker than that, and you have up-to-date information (the cache file is a snapshot from some time ago).

Having said that, QDirStat also comes with a Perl script that can create a cache file as a background / cron job; thus, you can even easily use it for a remote headless server, fetch that information and view it at your leisure on your desktop machine with QDirStat. See also

- https://github.com/shundhammer/qdirstat/blob/master/scripts/README.md
- https://github.com/shundhammer/qdirstat/blob/master/doc/QDirStat-for-Servers.md
- https://github.com/shundhammer/qdirstat/blob/master/doc/cache-file-format.txt

HTH

**Comment by shundhammer** [comment]
As for the "move to trash" feature, feel free to lift code from or be inspired by the QDirStat implementation:

https://github.com/shundhammer/qdirstat/blob/master/src/Trash.h
https://github.com/shundhammer/qdirstat/blob/master/src/Trash.cpp

**Comment by shundhammer** [comment]
For your use case, maybe consider periodically regenerating cache files and dropping them in the subtree that they describe, so QDirStat will pick them up as it finds them and use them instead of descending into a subtree. That is intended for large subtrees with less frequent changes; maybe it turns out to be useful for you.

https://github.com/shundhammer/qdirstat/blob/master/scripts/README.md

> For large directories (archives etc.) that don't change that much, you can also generate a QDirStat cache file (either with the Perl script or with QDirStat itself) and save it to that corresponding directory.
>
> If QDirStat finds a file .qdirstat.cache.gz in a directory, it checks if the toplevel directory in that cache file is the same as the current directory, and if it is, it uses the cache file for that directory rather than reading all subdirectories from disk. If you or the users of that machine use QDirStat often, this might take a lot of I/O load from the server.

**Comment by aspiers** [comment]
@shundhammer Thanks a lot for all the very helpful info!!

**Comment by l8gravely** [comment]
At this point, unless someone else implements this feature and proposes a patch, I'm closing this issue.  Thanks for the great discussion.
